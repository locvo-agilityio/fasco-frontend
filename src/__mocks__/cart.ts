import { ICart } from '@/types';

export const MOCK_CARTS: ICart[] = [
  {
    id: '1',
    name: 'Sofa',
    quantity: 1,
    price: 9550,
    image:
      'https://down-vn.img.susercontent.com/file/vn-11134207-7ras8-m3sbmx26g0cc16.webp',
    color: 'White',
    totalQuantity: 5,
    size: 'M',
    total: 9550,
    productId: '',
  },
  {
    id: '2',
    name: 'Sofa',
    quantity: 1,
    price: 9550,
    image:
      'https://down-vn.img.susercontent.com/file/vn-11134207-7ras8-m3sbmx26g0cc16.webp',
    color: 'Red',
    totalQuantity: 5,
    size: 'L',
    total: 9550,
    productId: '',
  },
];
