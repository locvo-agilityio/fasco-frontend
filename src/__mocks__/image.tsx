export const MOCK_IMAGE_GALLERY = [
  {
    id: 1,
    image:
      'https://down-vn.img.susercontent.com/file/sg-11134201-7rdxx-lz73hliz42dj59.webp',
  },
  {
    id: 2,
    image:
      'https://down-vn.img.susercontent.com/file/sg-11134301-7rdxl-lydeljvb148a71.webp',
  },
  {
    id: 3,
    image:
      'https://down-vn.img.susercontent.com/file/sg-11134301-7rdwt-lydeljvb144o82.webp',
  },
  {
    id: 4,
    image:
      'https://down-vn.img.susercontent.com/file/sg-11134301-7rd70-lu7m75qrwqihda.webp',
  },
  {
    id: 5,
    image:
      'https://down-vn.img.susercontent.com/file/sg-11134301-7rd6a-lu7m75szs3i1cc.webp',
  },
  {
    id: 6,
    image:
      'https://down-vn.img.susercontent.com/file/sg-11134301-7rd4m-lu7m75sfswqn05.webp',
  },
  {
    id: 7,
    image:
      'https://down-vn.img.susercontent.com/file/sg-11134301-7rd73-lu7m75tjr9rted.webp',
  },
];
