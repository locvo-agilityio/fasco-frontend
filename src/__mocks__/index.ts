export * from './next-auth';
export * from './next-auth-providers-credentials';
export * from './image';
export * from './cart';
