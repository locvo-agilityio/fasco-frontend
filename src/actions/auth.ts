'use server';

import { AuthError } from 'next-auth';
import { redirect } from 'next/navigation';

// Services
import { apiClient } from '@/services';

// Models
import { IUserRequest } from '@/models';

// Constants
import {
  API_ENDPOINT,
  AUTH_ERROR_TYPES,
  AUTH_METHOD,
  ERROR_MESSAGES,
  ROUTES,
  SUCCESS_MESSAGES,
  ToastType,
} from '@/constants';

// Auth
import { signIn } from '@/auth';

const signUp = async (_: { message: string }, data: FormData) => {
  let errorOccurred = false;

  try {
    const payload = {
      email: data.get('email'),
      password: data.get('password'),
      firstName: data.get('firstName'),
      lastName: data.get('lastName'),
      username: data.get('email'),
    };

    await apiClient.post<IUserRequest>(API_ENDPOINT.SIGN_UP, payload);

    return {
      status: ToastType.SUCCESS,
      message: SUCCESS_MESSAGES.SIGN_UP_SUCCESS,
    };
  } catch (error) {
    errorOccurred = true;
    return {
      status: ToastType.ERROR,
      message:
        error instanceof Error ? error.message : ERROR_MESSAGES.UNKNOWN_ERROR,
    };
  } finally {
    if (!errorOccurred) {
      await signIn(AUTH_METHOD.CREDENTIALS, {
        email: data.get('email'),
        password: data.get('password'),
        redirectTo: ROUTES.HOME,
      });
    }
  }
};

const login = async (_: { message: string }, data: FormData) => {
  let errorOccurred = false;

  try {
    await signIn(AUTH_METHOD.CREDENTIALS, data);

    return {
      status: ToastType.SUCCESS,
      message: SUCCESS_MESSAGES.SIGN_IN_SUCCESS,
    };
  } catch (error) {
    if (error instanceof AuthError) {
      errorOccurred = true;

      switch (error.type) {
        case AUTH_ERROR_TYPES.CREDENTIALS_SIGN_IN:
          return {
            status: ToastType.ERROR,
            message: ERROR_MESSAGES.EMAIL_PASSWORD_INVALID,
          };

        case AUTH_ERROR_TYPES.CALLBACK_ROUTE_ERROR:
          return {
            status: ToastType.ERROR,
            message: ERROR_MESSAGES.EMAIL_PASSWORD_INVALID,
          };

        default:
          return {
            status: ToastType.ERROR,
            message: ERROR_MESSAGES.UNKNOWN_ERROR,
          };
      }
    }

    throw error;
  } finally {
    if (!errorOccurred) redirect(ROUTES.HOME);
  }
};

const signInWithGitHub = async () => await signIn(AUTH_METHOD.GITHUB);

const signInWithGoogle = async () => await signIn(AUTH_METHOD.GOOGLE);

export { signUp, login, signInWithGitHub, signInWithGoogle };
