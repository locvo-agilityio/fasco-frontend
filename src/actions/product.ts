'use server';

import { revalidateTag } from 'next/cache';

// Services
import { apiClient } from '@/services';

// Constants
import { API_ENDPOINT, DEFAULT_PAGE, QUERY, TAG_KEYS } from '@/constants';

// Models
import { IProduct } from '@/models';

// Types
import { ICart, IPagination, ISearchParams } from '@/types';

// Utils
import { mergeProducts } from '@/utils';

export const getProductNewArrivals = async () => {
  const url = `${API_ENDPOINT.PRODUCTS}?${QUERY.NEW_ARRIVALS}`;

  return await apiClient.get<{ data: IProduct[] }>(url, {
    cache: 'force-cache', //Next.js 15 default is 'no-cache'
  });
};

export const getProducts = async (params: ISearchParams) => {
  const {
    size = '',
    color = '',
    collection = '',
    brand = '',
    price = '',
    page = DEFAULT_PAGE,
  } = params || {};

  const url = `${API_ENDPOINT.PRODUCTS}?${QUERY.PRODUCTS(
    size,
    color,
    brand,
    collection,
    price,
    page,
  )}`;

  return await apiClient.get<{
    data: IProduct[];
    meta: { pagination: IPagination };
  }>(url, {
    cache: 'force-cache', //Next.js 15 default is 'no-cache'
  });
};

export const getProduct = async (id: string) => {
  const url = `${API_ENDPOINT.PRODUCTS}/${id}?populate=*`;

  return (
    await apiClient.get<{ data: IProduct }>(url, {
      cache: 'force-cache',
      next: { tags: [TAG_KEYS.PRODUCT_BY_ID(id)] },
    })
  ).data;
};

export const checkoutCart = async (cart: ICart[]) => {
  const carts = mergeProducts(cart);

  return Promise.all(
    carts.map(async ({ productId, quantity: quantityCart }) => {
      const url = `${API_ENDPOINT.PRODUCTS}/${productId}?populate=*`;

      try {
        const product = (await apiClient.get<{ data: IProduct }>(url)).data;

        if (product) {
          const { quantity } = product.data;

          return { productId, quantity, quantityCart };
        }

        // eslint-disable-next-line @typescript-eslint/no-unused-vars
      } catch (error) {
        return { productId, quantity: 0, quantityCart };
      }
    }),
  );
};

export const updateQuantityProduct = async (
  id: string,
  quantity: number,
  quantityCart: number,
) => {
  await apiClient.put<{ data: IProduct }>(`${API_ENDPOINT.PRODUCTS}/${id}`, {
    data: { quantity: quantity - quantityCart },
  });

  revalidateTag(TAG_KEYS.PRODUCT_BY_ID(id));
};
