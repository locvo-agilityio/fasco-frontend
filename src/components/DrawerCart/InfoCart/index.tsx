import { memo, useCallback } from 'react';
import Image from 'next/image';
import isEqual from 'react-fast-compare';

// Utils
import { formatNumber } from '@/utils';

// Components
import NumberInput from '@/components/NumberInput';

// Types
import { ICart } from '@/types';

interface InfoCartProps {
  productId: string;
  image: string;
  name: string;
  color: string;
  price: number;
  minQuantity: number;
  totalQuantity: number;
  onChangeQuantity: ({ productId, quantity }: ICart) => void;
}

const InfoCart = ({
  productId,
  image,
  name,
  color,
  price,
  minQuantity,
  totalQuantity,
  onChangeQuantity,
}: InfoCartProps) => {
  const handleQuantity = useCallback(
    (value: number) => {
      onChangeQuantity({ productId, quantity: value } as ICart);
    },
    [onChangeQuantity, productId],
  );

  return (
    <div className="flex flex-row gap-5 pb-5 border-b-1 border-primary-100">
      <div className="w-[168px] h-[225px] border overflow-hidden relative">
        <Image
          src={image}
          alt={`Product ${name}`}
          fill
          sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
          priority
          className="object-cover"
        />
      </div>
      <div>
        <p className="font-volkhov text-2md">{name}</p>
        <p className="font-poppins text-2md text-primary-100">
          Color : {color}
        </p>
        <p className="font-poppins text-2md">${formatNumber(price)}</p>

        <div className="mt-2">
          <NumberInput
            min={minQuantity}
            max={totalQuantity}
            onChange={handleQuantity}
          />
        </div>
      </div>
    </div>
  );
};

export default memo(InfoCart, isEqual);
