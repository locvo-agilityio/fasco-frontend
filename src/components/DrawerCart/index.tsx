'use client';

import React, { useTransition } from 'react';
import {
  Drawer,
  DrawerContent,
  DrawerHeader,
  DrawerBody,
  DrawerFooter,
  useDisclosure,
  Checkbox,
  addToast,
} from '@heroui/react';
import { ShoppingBagIcon } from 'lucide-react';
import Link from 'next/link';

// Components
import { Button } from '@/components/common';
import InfoCart from './InfoCart';

// Utils
import { formatNumber, totalPrice } from '@/utils';

// Context
import { useCart } from '@/contexts';

// Constants
import {
  ERROR_MESSAGES,
  ROUTES,
  SUCCESS_MESSAGES,
  ToastType,
} from '@/constants';

// Actions
import { checkoutCart, updateQuantityProduct } from '@/actions';

const DrawerCart = () => {
  const { cart, updateCart, removeCart } = useCart();
  const [isPending, startTransition] = useTransition();
  const { isOpen, onOpen, onClose } = useDisclosure();

  const handleCheckoutCart = async () => {
    try {
      startTransition(async () => {
        const carts = await checkoutCart(cart);

        await Promise.all(
          carts.filter(Boolean).map(async (item) => {
            if (item?.productId) {
              await updateQuantityProduct(
                item.productId,
                item.quantity,
                item.quantityCart,
              );
            }
          }),
        );

        startTransition(() => {
          removeCart();
          onClose();

          addToast({
            description: SUCCESS_MESSAGES.CHECKOUT_SUCCESS,
            color: ToastType.SUCCESS,
          });
        });
      });
    } catch (error) {
      addToast({
        description:
          error instanceof Error ? error.message : ERROR_MESSAGES.UNKNOWN_ERROR,
        color: 'danger',
      });
    }
  };

  return (
    <div>
      <div className="relative">
        <Button
          title="Cart"
          variant="ghost"
          size="md"
          color="primary"
          className="w-5 min-w-5 px-0 py-0 border-none data-[hover=true]:!bg-transparent data-[hover=true]:!text-primary-foreground"
          onPress={onOpen}
        >
          <ShoppingBagIcon className="text-primary-200" size={18} />
        </Button>
        {Boolean(cart.length) && (
          <span className="absolute top-0 left-3 flex items-center justify-center w-4 h-4 text-xs font-semibold text-white bg-danger-300 rounded-full">
            {cart.length}
          </span>
        )}
      </div>
      <Drawer
        isOpen={isOpen}
        size="3xl"
        onClose={onClose}
        className="rounded-none"
      >
        <DrawerContent>
          {() => (
            <>
              <DrawerHeader className="flex flex-col gap-6 items-start font-volkhov text-3xl font-normal">
                Shopping Cart
                <p className="font-poppins text-3md text-primary-100">
                  Buy <span className="font-bold text-black">$122.35</span> More
                  And Get{' '}
                  <span className="font-bold text-black">Free Shipping</span>
                </p>
              </DrawerHeader>
              <DrawerBody>
                {Boolean(cart.length) ? (
                  cart.map(
                    ({
                      productId,
                      image,
                      name,
                      color,
                      quantity: minQuantity,
                      size,
                      price,
                      priceSale,
                      totalQuantity,
                    }) => {
                      const keyCart = `cart-${productId}-${color}-${size}`;

                      return (
                        <InfoCart
                          key={keyCart}
                          productId={productId}
                          image={image}
                          name={name}
                          color={color}
                          minQuantity={minQuantity}
                          price={priceSale || price}
                          totalQuantity={totalQuantity}
                          onChangeQuantity={updateCart}
                        />
                      );
                    },
                  )
                ) : (
                  <div className="w-full h-full flex items-center justify-center">
                    <p className="font-volkhov text-2md text-center text-black">
                      {ERROR_MESSAGES.EMPTY_CART}
                    </p>
                  </div>
                )}
              </DrawerBody>
              <DrawerFooter className="flex flex-col gap-4">
                <Checkbox
                  radius="none"
                  color="default"
                  className="w-full max-w-3xl pb-5 mt-5 border-b-1 border-primary-100"
                >
                  <p className="font-poppins text-2md text-primary-100">
                    For <span className="text-black">$10.00</span> please wrap
                    the product
                  </p>
                </Checkbox>

                <div className="w-full mt-11 flex justify-between font-volkhov text-2md">
                  <p>Subtotal</p>
                  <p>${formatNumber(totalPrice(cart))}</p>
                </div>

                <Button
                  onPress={handleCheckoutCart}
                  className="h-16 disabled:cursor-not-allowed"
                  isDisabled={!cart.length || isPending}
                  isLoading={isPending}
                >
                  Checkout
                </Button>
                <Button
                  as={Link}
                  href={ROUTES.CART}
                  color="primary"
                  variant="ghost"
                  size="md"
                  className="border-none text-black underline data-[hover=true]:!bg-transparent data-[hover=true]:!text-blue-500"
                >
                  View Cart
                </Button>
              </DrawerFooter>
            </>
          )}
        </DrawerContent>
      </Drawer>
    </div>
  );
};

export default DrawerCart;
