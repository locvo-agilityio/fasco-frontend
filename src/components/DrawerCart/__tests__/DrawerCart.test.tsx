import { fireEvent, render, waitFor } from '@testing-library/react';

// Components
import DrawerCart from '..';

// Mocks
import { MOCK_CARTS } from '@/__mocks__';

// Actions
import { checkoutCart } from '@/actions';

jest.mock('@/contexts', () => ({
  useCart: jest.fn(() => ({
    cart: MOCK_CARTS,
    updateCart: jest.fn(),
    removeCart: jest.fn(),
  })),
}));

jest.mock('@heroui/react', () => ({
  ...jest.requireActual('@heroui/react'),
  useDisclosure: jest.fn(() => ({
    onOpen: jest.fn(),
    onClose: jest.fn(),
    isOpen: true,
  })),
}));

jest.mock('@/actions', () => ({
  checkoutCart: jest.fn(() => Promise.resolve(MOCK_CARTS)),
  updateQuantityProduct: jest.fn(() => Promise.resolve({})),
}));

describe('DrawerCart component', () => {
  beforeAll(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { container } = render(<DrawerCart />);
    expect(container).toMatchSnapshot();
  });

  it('handle click checkout', async () => {
    const { getByRole } = render(<DrawerCart />);

    const checkoutButton = getByRole('button', { name: 'Checkout' });

    await waitFor(() => {
      fireEvent.click(checkoutButton);
    });

    await waitFor(() => {
      expect(checkoutCart).toHaveBeenCalled();
    });
  });
});
