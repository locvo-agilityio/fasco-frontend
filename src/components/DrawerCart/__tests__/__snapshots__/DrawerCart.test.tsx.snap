// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerCart component should render correctly 1`] = `
<div
  aria-hidden="true"
>
  <div>
    <div
      class="relative"
    >
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 h-10 gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent data-[hover=true]:!border-none text-secondary-400 border border-secondary-100 disabled:opacity-50 text-base rounded-base data-[pressed=true]:scale-[1] font-poppins w-5 min-w-5 px-0 py-0 border-none data-[hover=true]:!bg-transparent data-[hover=true]:!text-primary-foreground"
        title="Cart"
        type="button"
      >
        <svg
          class="lucide lucide-shopping-bag text-primary-200"
          fill="none"
          height="18"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 2 3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4Z"
          />
          <path
            d="M3 6h18"
          />
          <path
            d="M16 10a4 4 0 0 1-8 0"
          />
        </svg>
      </button>
      <span
        class="absolute top-0 left-3 flex items-center justify-center w-4 h-4 text-xs font-semibold text-white bg-danger-300 rounded-full"
      >
        2
      </span>
    </div>
  </div>
</div>
`;
